File: ios/Bridge/Hybrid/HybridTextRecognizer.swift
import Foundation
import NitroModules
import VisionCamera

final class HybridTextRecognizer: HybridTextRecognizerSpec {
  private let options: TextRecognizerOptions
  private let recognitionQueue = DispatchQueue(
    label: "com.visioncameramlkit.text-recognizer"
  )
  #if MLKIT_TEXT_RECOGNITION_ANY
    private let imagePreprocessor = ImagePreprocessor()
    private let textOptions: TextRecognitionOptions
    private let imageOptions: ImagePreprocessingOptions
    private let recognitionService: MLKitTextRecognitionService
    private let staticRecognitionService: MLKitTextRecognitionService
  #endif

  init(options: TextRecognizerOptions) throws {
    self.options = options
    #if MLKIT_TEXT_RECOGNITION_ANY
      let resolvedTextOptions = options.toDomainTextRecognitionOptions()
      self.textOptions = resolvedTextOptions
      self.imageOptions = options.toImagePreprocessingOptions()
      self.recognitionService = try TextRecognitionServiceFactory.create(
        language: resolvedTextOptions.language
      )
      self.staticRecognitionService = try TextRecognitionServiceFactory.create(
        language: resolvedTextOptions.language
      )
    #endif
  }

  func recognize(frame: any HybridFrameSpec, args: MLKitBaseArguments?) throws
    -> TextRecognitionResult
  {
    #if MLKIT_TEXT_RECOGNITION_ANY
      let callImageOptions = imageOptions.withOutputOrientation(
        args?.outputOrientation?.toOutputOrientation()
      )
      guard
        let processedImage = imagePreprocessor.preprocessFrame(
          frame: frame,
          options: callImageOptions
        )
      else {
        throw RuntimeError.error(withMessage: "Failed to preprocess VisionCamera frame.")
      }

      let result = try recognitionService.recognize(image: processedImage)
      return result.remapped(using: processedImage.metadata).toNitroResult()
    #else
      throw RuntimeError.error(
        withMessage: "TextRecognition is not enabled in the native ML Kit configuration.")
    #endif
  }

  func recognizeImage(uri: String) throws -> Promise<TextRecognitionResult> {
    #if MLKIT_TEXT_RECOGNITION_ANY
      return Promise.parallel(recognitionQueue) {
        let imageFile = try StaticImageURLResolver.resolve(uri)
        guard
          let processedImage = self.imagePreprocessor.preprocessImage(
            imageFile: imageFile,
            options: self.imageOptions
          )
        else {
          throw RuntimeError.error(withMessage: "Failed to preprocess static image: \(uri)")
        }

        let result = try self.staticRecognitionService.recognize(image: processedImage)
        return result.remapped(using: processedImage.metadata).toNitroResult()
      }
    #else
      return Promise.rejected(
        withError: RuntimeError.error(
          withMessage: "TextRecognition is not enabled in the native ML Kit configuration."
        )
      )
    #endif
  }
}

File: android/src/main/java/com/margelo/nitro/visioncameramlkit/HybridTextRecognizer.kt
package com.margelo.nitro.visioncameramlkit

import android.util.Log
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.camera.HybridFrameSpec
import com.margelo.nitro.core.Promise
import com.visioncameramlkit.domain.models.remap
import com.visioncameramlkit.infrastructure.image.ImagePreprocessor
import com.visioncameramlkit.infrastructure.mlkit.factories.TextRecognitionServiceFactory
import java.util.concurrent.atomic.AtomicBoolean

@DoNotStrip
@Keep
class HybridTextRecognizer(
  private val options: TextRecognizerOptions,
) : HybridTextRecognizerSpec() {
  private val imagePreprocessor = ImagePreprocessor()
  private val textOptions = options.toDomainTextRecognitionOptions()
  private val imageOptions = options.toImagePreprocessingOptions()
  private val recognitionService = TextRecognitionServiceFactory.create(textOptions.language)
  private val isDisposed = AtomicBoolean(false)

  override fun recognize(
    frame: HybridFrameSpec,
    args: MLKitBaseArguments?,
  ): TextRecognitionResult {
    val processedImage = frame.toProcessedImage(imagePreprocessor, imageOptions)
    val result = recognitionService.recognize(processedImage)
    return result.remap(processedImage.metadata).toNitroResult()
  }

  override fun recognizeImage(uri: String): Promise<TextRecognitionResult> =
    Promise.parallel {
      StaticImageUriResolver.resolve(uri).use { resolvedImage ->
        val processedImage = imagePreprocessor.preprocessImage(resolvedImage.file, imageOptions)
        val result = recognitionService.recognize(processedImage)
        result.remap(processedImage.metadata).toNitroResult()
      }
    }

  override fun dispose() {
    if (isDisposed.compareAndSet(false, true)) {
      runCatching { recognitionService.close() }
        .onFailure { error -> Log.w(TAG, "Failed to close text recognizer", error) }
    }
    super.dispose()
  }
}
